### DISCUSSION TITLE
classes/class_timer

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#170] [User-contributed notes]
# classes/class_timer

Inherits: Node< Object A countdown timer. Description: The Timer node is a countdown timer and is the simplest way to handle time-based logic in the engine. When a timer reaches the end of its wait...

https://docs.godotengine.org/en/stable/classes/class_timer.html

**Comment by nziswat** [comment]
>Starts the timer, if it was not started already.

The above line seems to be incorrect, from what I have seen calling start on a timer that has already started will restart it, instead of just doing nothing as the line implies.

**Reply by jessodum** [reply]
Thank you. I was curious about whether this would be the case or not.

**Reply by Sugo98** [reply]
Thank you. I was searching exactly this

**Reply by Calinou** [reply]
This was fixed in the class reference in August, but hasn't been cherry-picked to the `4.3` branch yet:

- https://github.com/godotengine/godot/pull/95485

**Comment by PPrinsloo** [comment]
Us there a way to get "time left" as milliseconds?

**Reply by Calinou** [reply]
Yes, `time_left * 1000` should do the trick.

**Reply by PPrinsloo** [reply]
This will simply return the time in increments of 1000. I need it to return milliseconds as milliseconds pass because I need more granular time. 

I have gone with a solution using delta time on the process method which is working. I think using the Timer was not the correct approach to begin with.

**Reply by Calinou** [reply]
`time_left` is a float that represents seconds, so multiplying it by `1000` will give you the value in milliseconds down to the nearest frame. (This means its precision depends on framerate, since timers only process on every process tick or physics tick depending on configuration.)

If you want to know engine time in milliseconds accurately (since the engine has started), use `Time.get_ticks_msec()` or for even more precision, `TIme.get_ticks_usec()`. These are also guaranteed to be monotonic, i.e. they will never decrease over time even if the user changes their system time or DST kicks in.

**Reply by PPrinsloo** [reply]
Ah nice, did not know that. Thanks!

**Comment by SerhiiHorbanov** [comment]
It would be very heplfull if they specified whether a timer is still running while outside tree

**Comment by Dharhix** [comment]
I'm not sure if this might be helpful to others, but I find myself in need of a way to trigger a timer before it's reached its ```wait_time```. Since ```time_left``` is read only I can't use that. I know this is very edge case, but a ```force_trigger()``` method to actually force the timer to reach its end (and restart in my case) would be welcomed.

**Reply by Calinou** [reply]
Feature proposals should be made on the [Godot proposals](https://github.com/godotengine/godot-proposals) repository.

**Reply by DFGameDev** [reply]
A useful built-in function but you can do the same thing with a side function:

func ForceTimerTimeout():
	timer.stop()
	timer.emit_signal("timeout")

**Comment by brekreney** [comment]
You should clarify if running stop() emits timeout().

**Reply by Calinou** [reply]
Pull request opened:

- https://github.com/godotengine/godot/pull/107228

**Comment by UnTabouret** [comment]
I think the documentation for the `start` function should explicitly state that it overwrites `wait_time`. The current wording implies that it merely overrides it for the current run.
